File: Demo/ObjcTest/ObjcTest/ViewController.m
//
//  ViewController.m
//  ObjcTest
//
//  Created by Nicholas Della Valle on 10/24/22.
//

#import "ViewController.h"
@import VideeoSDK;

@interface ViewController () <VideeoStreamDelegate>

@property (nonatomic) UIButton *startButton;


@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.

    VideeoConfig *config = [[VideeoConfig alloc] initWithShopID:@"jaredvideeo"
                                                         locale:@"en"
                                                       currency:@"usd"
                                                    environment:VideeoEnvironmentTest
                                 allowsPictureInPicturePlayback:YES
                                                 showCartButton:YES];

    VideeoUser *user = [[VideeoUser alloc] initWithFirstName:@"Jared"
                                                    lastName:@"Green"
                                                  profileURL:@"https://picsum.photos/200"
                                                  facebookId:@"123"
                                                 instagramId:nil
                                                    tiktokId:nil];
   
    [[VideeoManager instance] initializeWithConfig:config videeoUser:user];
    
    __weak ViewController *weakSelf = self;
    [[NSNotificationCenter defaultCenter] addObserverForName:NSNotification.videeoLiveStatusChanged object:[VideeoManager instance] queue:nil usingBlock:^(NSNotification* notification){
        
        if ([notification.userInfo[@"isLive"] boolValue] == true) {
            [weakSelf alertShowLive];
        }
    }];
    
}

- (void)alertShowLive {
    
    if ([self presentedViewController] != nil) {
        return;
    }
    
    UIAlertController *alertVC = [UIAlertController alertControllerWithTitle:@"Watch live stream?" message:nil preferredStyle:UIAlertControllerStyleAlert];
    [alertVC addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    [alertVC addAction:[UIAlertAction actionWithTitle:@"Watch" style:UIAlertActionStyleDefault  handler:^(UIAlertAction *action){
        [self showLive: nil];
    }]];
    
    [self presentViewController:alertVC animated:true completion:nil];
    
}

- (IBAction)showLive:(id)sender {
    NSError *error;
    UIViewController *vc = [[VideeoManager instance] getLiveStreamViewControllerWithDelegate:self error:&error];

    if (error == nil) {
        vc.modalPresentationStyle = UIModalPresentationFullScreen;
        [self presentViewController:vc animated:true completion:nil];
    } else {
        NSLog(@"%@", error.localizedDescription);
    }
}

- (IBAction)showReplays:(id)sender {
    //TODO: How do we handle replays now that the Videeo SDK does not include any replay UI?
}

- (void)productTappedWithProduct:(id<VideeoProduct> _Nonnull)product presenter:(id<ProductPresenter> _Nonnull)presenter {}

- (void)cartTappedWithPresenter:(id<CartPresenter> _Nonnull)presenter {}

- (UIViewController * _Nullable)cartTapped {
    return nil;
}

- (void)close {
    [self dismissViewControllerAnimated:true completion:nil];
}

- (void)userAuthenticationRequired {}

- (BOOL)streamEnded {
    return false;
}

- (BOOL)userRemovedFromLive {
    return false;
}

@end

File: Demo/UIKit/Demo/ViewController.swift
//
//  ViewController.swift
//  Demo
//
//  Created by Jared Green on 10/17/22.
//

import UIKit
import VideeoSDK


class ViewController: UIViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        //Configure the Videeo Manager with the shop name, environment and user.
        //The user parameter is option in the initialize function and can be supplied
        //directly to the VideeoManager at any point in the app lifecycle by calling setVideeoUser.
        let config = VideeoConfig(shopID: "jaredvideeo",
                                  locale: "en",
                                  currency: "USD",
                                  environment: .test)
        let videeoUser = VideeoUser(firstName: "Jared",
                                    lastName: "Green",
                                    profileURL: "https://picsum.photos/200",
                                    facebookId: "123",
                                    instagramId: "456",
                                    tiktokId: "789")
        VideeoManager.instance.initialize(config: config, videeoUser: videeoUser)
        
        //Register for videeo status events so the app will be notifed if a live stream starts or stops.
        NotificationCenter.default.addObserver(
            forName: .videeoLiveStatusChanged,
            object: VideeoManager.instance,
            queue: nil) { [weak self] notification in
                if let userInfo = notification.userInfo, let isLive = userInfo["isLive"] as? Bool, isLive {
                    self?.alertStreamIsLive()
                }
            }
    }
    
    @IBAction func showLive(_ sender: Any) {
        showLiveStream()
    }

    @IBAction func showReplays(_ sender: Any) {
        showReplays()
    }

    private func alertStreamIsLive() {
        guard self.presentedViewController == nil else { return }
        
        let okAction = UIAlertAction(title: "OK", style: .default) { [weak self] _ in
            self?.showLiveStream()
        }
        
        let alertController = UIAlertController(title: "Watch live stream?", message: nil, preferredStyle: .alert)
        alertController.addAction(okAction)
        alertController.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        
        present(alertController, animated: true)
    }
    
    private func showLiveStream() {
        if let videeoVC = try? VideeoManager.instance.getLiveStreamViewController(delegate: self) {
            videeoVC.modalPresentationStyle = .fullScreen
            present(videeoVC, animated: true)
        }
    }

    private func showReplays() {
        //TODO: How do we handle replays now that the Videeo SDK does not include any replay UI?
    }

}

extension ViewController: VideeoStreamDelegate {
    func productTapped(product: VideeoProduct, presenter: ProductPresenter) {}

    func cartTapped(presenter: CartPresenter) {}

    func close() {
        dismiss(animated: true)
    }

    func userAuthenticationRequired() {}

    func streamEnded() -> Bool {
        return false
    }

    func userRemovedFromLive() -> Bool {
        return false
    }
}

